Тема “Обучение с учителем”


Задание 1

- Импортируйте библиотеки pandas и numpy.

- Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn.

- Создайте датафреймы X и y из этих данных.

- Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так,

чтобы размер тестовой выборки составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.

- Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.

- Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.

- Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.


In [1]:
import numpy as np
import pandas as pd

In [2]:
from sklearn.datasets import load_boston

In [19]:
boston = load_boston()
data = boston.data

In [14]:
boston.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

In [22]:
feature_names = boston['feature_names']
x = pd.DataFrame(data, columns=feature_names)
x.head()
target = boston['target']
y = pd.DataFrame(target, columns=['price'])
y.head()

price
0   24.0
1   21.6
2   34.7
3   33.4
4   36.2

In [23]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)

In [17]:
from sklearn.linear_model import LinearRegression

In [18]:
lr = LinearRegression()

In [24]:
lr.fit(x_train, y_train)

LinearRegression()

In [28]:
y_pred = lr.predict(x_test)
check_test = pd.DataFrame({
    'y_test': y_test['price'], 
    'y_pred_lr': y_pred.flatten()})
check_test.head()

y_test  y_pred_lr
173    23.6  28.648960
274    32.4  36.495014
491    13.6  15.411193
72     22.8  25.403213
452    16.1  18.855280

In [33]:
from sklearn.metrics import r2_score

In [34]:
r2_score(y_test, y_pred)

0.711226005748496

---

Задание 2

- Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.

- Сделайте агрумент n_estimators равным 1000,
max_depth должен быть равен 12 и random_state сделайте равным 42.

- Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression,
но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0],
чтобы получить из датафрейма одномерный массив Numpy,
так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно 
применение массивов вместо датафрейма.

- Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего 
задания.

- Напишите в комментариях к коду, какая модель в данном случае работает лучше.


In [35]:
from sklearn.ensemble import RandomForestRegressor

In [36]:
clf = RandomForestRegressor(n_estimators=1000, max_depth=12, random_state=42)

In [37]:
clf.fit(x_train, y_train.values[:, 0])

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

In [38]:
y_pred_clf = clf.predict(x_test)
check_test_clf = pd.DataFrame({'y_test': y_test['price'],
                               'y_pred_clf': y_pred_clf.flatten()})
check_test_clf.head()

y_test  y_pred_clf
173    23.6   22.806412
274    32.4   31.131464
491    13.6   16.339125
72     22.8   23.810726
452    16.1   17.139521

In [39]:
r2_score(y_test, y_pred_clf)

0.87472606157312

* r2_score у модели линейной регрессии = 0.711226005748496
* r2_score у случайного леса = 0.87472606157312

ВЫВОД: 0.87 ближе к 1 => случайный лес работает лучше.

---


*Задание 3

* Вызовите документацию для класса RandomForestRegressor,
найдите информацию об атрибуте feature_importances_.

* С помощью этого атрибута найдите сумму всех показателей важности,
установите, какие два признака показывают наибольшую важность.


In [40]:
clf.feature_importances_

array([0.03167574, 0.00154252, 0.00713813, 0.00123624, 0.01426897,
       0.40268179, 0.01429864, 0.06397257, 0.00528122, 0.01152493,
       0.01808108, 0.01245085, 0.41584732])

In [41]:
feature_importance = pd.DataFrame({'name': x.columns, 
                                   'feature_importance': clf.feature_importances_}, 
                                  columns=['feature_importance', 'name'])
feature_importance

feature_importance     name
0             0.031676     CRIM
1             0.001543       ZN
2             0.007138    INDUS
3             0.001236     CHAS
4             0.014269      NOX
5             0.402682       RM
6             0.014299      AGE
7             0.063973      DIS
8             0.005281      RAD
9             0.011525      TAX
10            0.018081  PTRATIO
11            0.012451        B
12            0.415847    LSTAT

In [42]:
feature_importance.nlargest(2, 'feature_importance')

feature_importance   name
12            0.415847  LSTAT
5             0.402682     RM

---